### ISSUE TITLE
How to visualize the test results?

### ISSUE THREAD

**Issue by zeng-ziyin** [Matrix-ASC/DeLA#1]
Dear Authors:
How can we visualize the test results?

**Comment by binjiechen** [comment]
Hi, visualization relies on coordinates and colors, so you may follow these general steps:
1. Save paired point coordinates, colors, labels and predictions
These can be done through various ways, including:
① Pick a point cloud in the test set and predict it. You may refer to dataset get_test_item and test script.
② Modify dataset get_test_item to let it return full coordinates and  original colors, then save with predictions and labels (which is cum.argmax(dim=1) and y in test.py).
③ Pre construct a test set list and pass it to dataset, like scannetv2_val.txt in ScanNet, so the list order is constant, then save predictions only. The rest can be loaded like in get_item.
2. Convert predictions and labels to colors
Now say you have a prediction pred of shape N and a color table ctab of shape C x 3. (classes by rgb)
You can do the conversion by col_pred = ctab[pred.long()]
3. Use a library to visualize the coordiate color pair
There are many libraries. Taking vedo for example.
```Python
import torch
from vedo import *

# input xyz of N x 3, float
# input col of N x 3, uint8
# add alpha
col = torch.cat([col, torch.full((col.shape[0], 1), 255, dtype=torch.uint8)], dim=1)
pts = Points(xyz.numpy(), r=4, c=col.float().numpy())
plt = Plotter(shape=(1,1), interactive=False, screensize=[2560,1440], size=[2560,1440])
plt.show(pts, at=0, resetcam=True)
# For another, change shape too
# plt.show(pts2, at=1, resetcam=True)

plt.interactor.Start()
```

**Comment by zouyanmei** [comment]
`
import torch
from torch import nn
import torch.nn.functional as F
from vision_s3dis import S3DIS, s3dis_test_collate_fn
from torch.utils.data import DataLoader
import os
import sys
from pathlib import Path
sys.path.append(str(Path(__file__).absolute().parent.parent))
import utils.util as util
from delasemseg import DelaSemSeg
from config import s3dis_args, dela_args
from torch.cuda.amp import autocast
import numpy as np
import random
def setup_seed(seed):
     torch.manual_seed(seed)
     torch.cuda.manual_seed_all(seed)
     np.random.seed(seed)
     random.seed(seed)
     torch.backends.cudnn.deterministic = True
setup_seed(20)

def write_obj(points, colors, out_filename):
    N = points.shape[0]
    fout = open(out_filename, 'w')
    for i in range(N):
        c = colors[i]
        fout.write('v %f %f %f %f %f %f\n' % (points[i, 0], points[i, 1], points[i, 2], c[0], c[1], c[2]))
    fout.close()

label2color = {'0':     [0, 255, 0],
                   '1':       [0, 0, 255],
                   '2':        [0, 255, 255],
                   '3':        [255, 255, 0],
                   '4':      [255, 0, 255],
                   '5':      [100, 100, 255],
                   '6':        [200, 200, 100],
                   '7':       [170, 120, 200],
                   '8':       [255, 0, 0],
                   '9':        [200, 100, 100],
                   '10':    [10, 200, 100],
                   '11':       [200, 200, 200],
                   '12':     [50, 50, 50]}

def write_obj_label(points, label, out_filename):
    N = points.shape[0]
    fout = open(out_filename, 'w')
    for i in range(N):
        c = label2color[str(label[i])]
        fout.write('v %f %f %f %f %f %f\n' % (points[i, 0], points[i, 1], points[i, 2], c[0], c[1], c[2]))
    fout.close()

prefix_names = ['input-Area5-', 'gt-Area5-', 'pred-Area5-']

torch.set_float32_matmul_precision("high")

loop = 12

testdlr = DataLoader(S3DIS(s3dis_args, partition="5", loop=loop, train=False, test=True), batch_size=1,
                      collate_fn=s3dis_test_collate_fn, pin_memory=True, num_workers=8)

model = DelaSemSeg(dela_args).cuda()

util.load_state("pretrained/best.pt", model=model)

model.eval()

metric = util.Metric(13)
cum = 0
cnt = 0

save_dir = 'pretrained/s3dis/visualization/'
os.makedirs(save_dir, exist_ok=True) 
with torch.no_grad():
    idx = 0
    for xyz, feature, indices, nn, y, full_xyz_raw, full_col in testdlr:
            xyz = xyz.cuda(non_blocking=True)
            feature = feature.cuda(non_blocking=True)
            indices = [ii.cuda(non_blocking=True).long() for ii in indices[::-1]]
            nn = nn.cuda(non_blocking=True).long()
            with autocast():
                p = model(xyz, feature, indices)
            cum = cum + p[nn]
            cnt += 1
            if cnt % loop == 0:
                idx = idx +1
                y = y.cuda(non_blocking=True)
                metric.update(cum, y)
                #save_name1 = f'{save_dir}'+prefix_names[0]+str(idx)+".obj"
                #write_obj(full_xyz_raw, full_col, save_name1)
                #save_name2 = f'{save_dir}'+prefix_names[1]+str(idx)+".obj"
                #write_obj_label(full_xyz_raw, y.cpu().numpy(), save_name2)
                save_name3 = f'{save_dir}'+prefix_names[2]+str(idx)+".obj"
                write_obj_label(full_xyz_raw, cum.argmax(dim=1).cpu().numpy(), save_name3)
                cnt = cum = 0
                print(idx)

metric.print("test: ")

`
